### ISSUE TITLE
Get datetime template values in the user-configured timezone

### ISSUE THREAD

**Issue by tothandor** [apache/airflow#23400]
### Apache Airflow version

2.3.0 (latest released)

### What happened

One could expect by [the documentation about timezones](https://airflow.apache.org/docs/apache-airflow/stable/timezone.html), that Airflow will respect `default_timezone` (from airflow.cfg [core]) or a DAG's `timezone`, no matter how it stores timestamps internally (UTC).  
But the [template variables](https://airflow.apache.org/docs/apache-airflow/stable/templates-ref.html#variables) don't do that, which could result unexpected results or unnecessarily complicated expressions should be used all over.

### What you think should happen instead

Please provide timezone aware template variables! A `tz` suffix/prefix could denote the timezone aware variants. For example: `data_interval_start` => `data_interval_start_tz`, `ds` => `ds_tz`, `logical_date` => `logical_date_tz`.

### How to reproduce

For example I use local timezone (CET/CEST) in all of my DAGs which is ahead of GMT (UTC) by +1 or +2 hours in the summer (DST). Nearly all tasks have date arguments, and some of them are running between 01:00 and 02:00. For example for `execution_date="2022-05-01T01:30:00+02:00"`, this would mean, that the date will be **2022-05-01** in the winter, and **2022-04-30** in the summer in UTC.

Of course I could work this around by using `logical_date.in_tz(dag.timezone).date()` in every template, instead of just `ds`, but it isn't a really nice solution. (As a matter of fact I have rather patched taskinstance.py to have timezone aware template vars.)

### Operating System

CentOS Stream 8

### Deployment

Virtualenv installation

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by uranusjr** [comment]
No, all the datetimes are aware of timezones, they are simply all in UTC. There is a big difference between the two. In other words, the values specify the correct points of time, they simply don’t convert to the string values you expect.

Given that it’s too expensive to convert those datetimes in most situations (not to mention compatibility issues), the simplest solution would be to provide a template filter to convert to the default timezone, e.g. `{{ date_interval_start|user_tz }}`.
